pennylane/devices/qutrit_mixed/measure.py

```diff
@@ -0,0 +1,308 @@
+# Copyright 2018-2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Code relevant for performing measurements on a qutrit mixed state.
+"""
+
+from typing import Callable
+from string import ascii_letters as alphabet
+from pennylane import math
+from pennylane.ops import Sum, Hamiltonian
+from pennylane.measurements import (
+    StateMeasurement,
+    MeasurementProcess,
+    ExpectationMP,
+    StateMP,
+    ProbabilityMP,
+    VarianceMP,
+)
+from pennylane.operation import Observable
+from pennylane.typing import TensorLike
+
+from .utils import (
+    get_einsum_mapping,
+    resquare_state,
+    get_num_wires,
+    get_new_state_einsum_indices,
+    QUDIT_DIM,
+)
+from .apply_operation import apply_operation
+
+alphabet_array = math.asarray(list(alphabet))
+
+
+def apply_observable_einsum(obs: Observable, state, is_state_batched: bool = False):
+    r"""Applies an observable to a density matrix rho, giving obs@state
+
+    Args:
+        obs (Operator): Operator to apply to the quantum state
+        state (array[complex]): Input quantum state
+        is_state_batched (bool): Boolean representing whether the state is batched or not
+
+    Returns:
+        TensorLike: the result of obs@state
+    """
+
+    def map_indices(
```

Comment:
Similar comment, can something more general be defined for the mapping and then used for both the measurement and operation? (If not, to match the convention of the other file, this can be defined as outside the function).

Reply:
I think I may have over complicated this.
In `utils` there is a function called `get_einsum_mapping`:

```
def get_einsum_mapping(
    op, state, indices_mapping_function, is_state_batched):
   # find  indices of operator and state
   ...

  # apply mapping function
      return indices_mapping_function(
          state_indices=state_indices,
          kraus_index=kraus_index,
          row_indices=row_indices,
          new_row_indices=new_row_indices,
          col_indices=col_indices,
          new_col_indices=new_col_indices,
    )
```

I changed `map_indices` variable -> `indices_mapping_function` here for clarity.

`map_indices` here and `_map_indices_apply_channel` are inputted as `indices_mapping_function`'s value to find the indices mapping to put into einsum.

another option would be to make get_einsum_mapping to return a dictionary of the values and have `map_indices` here and `_map_indices_apply_channel`  call `get_einsum_mapping`  (or getting rid of these helpers entirely).

Reply:
Oh okay I see - I think for consistency then, this should be defined out of the scope of `apply_observable_einsum`, like is done with `_map_indices_apply_channel` in the `apply_operation.py` file (as @albi3ro suggests in the comment below)

Reply:
I agree, will add in next commit